// @op-engineering/op-sqlite/node/dist/database.js
import * as fs from 'node:fs';
import * as path from 'node:path';
import Database from 'better-sqlite3';
export class NodeDatabase {
    constructor(name, location) {
        this.dbPath = ':memory:';
        if (location !== ":memory:") {
            const dbLocation = location || './';
            this.dbPath = path.join(dbLocation, name);
            // Ensure directory exists
            const dir = path.dirname(this.dbPath);
            if (!fs.existsSync(dir)) {
                fs.mkdirSync(dir, { recursive: true });
            }
        }
        this.db = new Database(this.dbPath);
        this.setupHooks();
    }
    setupHooks() {
        // Setup update hook if needed
        // if (this.db.function) {
        // Note: better-sqlite3 doesn't have direct update hook support
        // This is a limitation compared to the native implementation
        // }
    }
    convertParams(params) {
        if (!params)
            return undefined;
        return params.map(param => {
            if (param instanceof ArrayBuffer) {
                return Buffer.from(param);
            }
            else if (ArrayBuffer.isView(param)) {
                return Buffer.from(param.buffer, param.byteOffset, param.byteLength);
            }
            return param;
        });
    }
    convertRows(stmt, rows) {
        const columnNames = stmt.columns().map(col => col.name);
        const metadata = stmt.columns().map((col, index) => ({
            name: col.name,
            type: col.type || 'UNKNOWN',
            index,
        }));
        return {
            rowsAffected: 0,
            rows: rows,
            columnNames,
            metadata,
        };
    }
    executeSync(query, params) {
        try {
            const convertedParams = this.convertParams(params);
            const stmt = this.db.prepare(query);
            // Check if it's a SELECT query
            const isSelect = query.trim().toUpperCase().startsWith('SELECT');
            if (isSelect) {
                const rows = convertedParams
                    ? stmt.all(...convertedParams)
                    : stmt.all();
                return this.convertRows(stmt, rows);
            }
            else {
                const info = convertedParams
                    ? stmt.run(...convertedParams)
                    : stmt.run();
                return {
                    rowsAffected: info.changes,
                    insertId: info.lastInsertRowid,
                    rows: [],
                };
            }
        }
        catch (error) {
            throw new Error(`SQL Error: ${error.message}`);
        }
    }
    async execute(query, params) {
        return Promise.resolve(this.executeSync(query, params));
    }
    async executeWithHostObjects(query, params) {
        // For Node.js, this is the same as execute since we don't have HostObjects
        return this.execute(query, params);
    }
    executeRawSync(query, params) {
        try {
            const convertedParams = this.convertParams(params);
            const stmt = this.db.prepare(query);
            const rows = convertedParams
                ? stmt.raw().all(...convertedParams)
                : stmt.raw().all();
            return rows;
        }
        catch (error) {
            throw new Error(`SQL Error: ${error.message}`);
        }
    }
    async executeRaw(query, params) {
        return Promise.resolve(this.executeRawSync(query, params));
    }
    async executeBatch(commands) {
        let totalRowsAffected = 0;
        const transaction = this.db.transaction(() => {
            for (const command of commands) {
                const [query, params] = command;
                if (Array.isArray(params) && params.length > 0 && Array.isArray(params[0])) {
                    // Multiple parameter sets
                    const stmt = this.db.prepare(query);
                    for (const paramSet of params) {
                        const converted = this.convertParams(paramSet);
                        const info = converted ? stmt.run(...converted) : stmt.run();
                        totalRowsAffected += info.changes;
                    }
                }
                else {
                    // Single parameter set or no parameters
                    const converted = this.convertParams(params);
                    const stmt = this.db.prepare(query);
                    const info = converted ? stmt.run(...converted) : stmt.run();
                    totalRowsAffected += info.changes;
                }
            }
        });
        transaction();
        return { rowsAffected: totalRowsAffected };
    }
    async loadFile(location) {
        const fileContent = fs.readFileSync(location, 'utf-8');
        const statements = fileContent
            .split(';')
            .map(s => s.trim())
            .filter(s => s.length > 0);
        let totalRowsAffected = 0;
        let commandCount = 0;
        const transaction = this.db.transaction(() => {
            for (const statement of statements) {
                const info = this.db.prepare(statement).run();
                totalRowsAffected += info.changes;
                commandCount++;
            }
        });
        transaction();
        return {
            rowsAffected: totalRowsAffected,
            commands: commandCount,
        };
    }
    async transaction(fn) {
        const transaction = {
            execute: async (query, params) => {
                return this.execute(query, params);
            },
            commit: async () => {
                return { rowsAffected: 0, rows: [] };
            },
            rollback: () => {
                throw new Error('ROLLBACK');
            },
        };
        // Manually control transaction with BEGIN/COMMIT/ROLLBACK to support async operations
        this.executeSync('BEGIN TRANSACTION');
        try {
            await fn(transaction);
            this.executeSync('COMMIT');
            if (this.commitHookCallback) {
                this.commitHookCallback();
            }
        }
        catch (error) {
            this.executeSync('ROLLBACK');
            if (this.rollbackHookCallback) {
                this.rollbackHookCallback();
            }
            throw error;
        }
    }
    prepareStatement(query) {
        const stmt = this.db.prepare(query);
        let boundParams = [];
        return {
            bindSync: (params) => {
                boundParams = this.convertParams(params) || [];
            },
            bind: async (params) => {
                boundParams = this.convertParams(params) || [];
            },
            execute: async () => {
                const isSelect = query.trim().toUpperCase().startsWith('SELECT');
                if (isSelect) {
                    const rows = boundParams.length > 0
                        ? stmt.all(...boundParams)
                        : stmt.all();
                    return this.convertRows(stmt, rows);
                }
                else {
                    const info = boundParams.length > 0
                        ? stmt.run(...boundParams)
                        : stmt.run();
                    return {
                        rowsAffected: info.changes,
                        insertId: info.lastInsertRowid,
                        rows: [],
                    };
                }
            },
        };
    }
    attach(params) {
        const dbLocation = params.location || './';
        const dbPath = path.join(dbLocation, params.secondaryDbFileName);
        this.db.prepare(`ATTACH DATABASE ? AS ?`).run(dbPath, params.alias);
    }
    detach(alias) {
        this.db.prepare(`DETACH DATABASE ?`).run(alias);
    }
    updateHook(callback) {
        this.updateHookCallback = callback;
        // Note: better-sqlite3 doesn't support update hooks directly
        console.warn('Update hooks are not fully supported in the Node.js implementation');
    }
    commitHook(callback) {
        this.commitHookCallback = callback;
    }
    rollbackHook(callback) {
        this.rollbackHookCallback = callback;
    }
    loadExtension(path, entryPoint) {
        // better-sqlite3 supports loadExtension but it may not be enabled by default
        try {
            if (entryPoint) {
                this.db.loadExtension(path, entryPoint);
            }
            else {
                this.db.loadExtension(path);
            }
        }
        catch (error) {
            throw new Error(`Failed to load extension: ${error.message}`);
        }
    }
    getDbPath(location) {
        return this.dbPath;
    }
    reactiveExecute(params) {
        // Reactive queries are not supported in Node.js implementation
        console.warn('Reactive queries are not supported in the Node.js implementation');
        return () => { };
    }
    sync() {
        // LibSQL sync is not supported in the Node.js implementation
        throw new Error('sync() is only available with libsql');
    }
    setReservedBytes(reservedBytes) {
        // SQLCipher specific, not supported in standard SQLite
        console.warn('setReservedBytes is not supported in the Node.js implementation');
    }
    getReservedBytes() {
        // SQLCipher specific, not supported in standard SQLite
        return 0;
    }
    async flushPendingReactiveQueries() {
        // No-op for Node.js implementation
    }
    close() {
        if (this.db && this.db.open) {
            this.db.close();
        }
    }
    delete(location) {
        this.close();
        const dbLocation = location || './';
        const dbPath = path.join(dbLocation, path.basename(this.dbPath));
        if (fs.existsSync(dbPath)) {
            fs.unlinkSync(dbPath);
        }
    }
}
